File: util/pwd.py
#!/usr/bin/env python3
import os
import os.path
import sys

def main():
    if len(sys.argv) >= 2:
        wd = sys.argv[1]
    else:
        wd = os.getcwd()
    home = os.path.expanduser("~")
    if wd.startswith(home):
        is_under_home = True
        wd = "~" + wd[len(home):]
    elif len(wd) > 0:
        is_under_home = False
        wd = wd[1:]
    if len(wd) >= 30:
        segs = wd.split("/")
        if len(segs) <= 2:
            omit_until = 0
        else:
            omit_until = len(segs) - 2
        for i in range(omit_until):
            if len(segs[i]) >= 1:
                segs[i] = segs[i][0]
        if not is_under_home:
            print("/", end="")
        print("/".join(segs), end="")
    else:
        print(wd, end="")

if __name__ == "__main__":
    main()

File: util/pwd.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cwd;

my $home = $ENV{"HOME"};
my $wd = Cwd::getcwd();
if ($#ARGV >= 0) {
    $wd = $ARGV[0];
}
my $is_under_home = 0;
if (index($wd, "^$home") == -1) {
    $is_under_home = 1;
    $wd =~ s/^$home/~/;
}
my $len = length($wd);
my @segs = split(/\//, $wd);
my $seglen = $#segs + 1;
my $index = 0;
while ($len > 30 && $index < $seglen - 2) {
    if (length($segs[$index]) > 1) {
	$len -= length($segs[$index]) - 1;
	$segs[$index] = substr($segs[$index], 0, 1);
    }
    $index++;
}
if (!$is_under_home) {
    print "/";
}
print join("/", @segs);
print "\n";
